torchaudio_contrib/augmentation.py

```diff
@@ -0,0 +1,151 @@
+import math
+import torch
+
+import torch.nn as nn
+from .layers import _ModuleNoStateBuffers, STFT
+from .functional import phase_vocoder
+
+
+class StretchSpecTime(_ModuleNoStateBuffers):
+    """
+    Stretch stft in time without modifying pitch for a given rate.
+
+    Args:
+
+        rate (float): rate to speed up or slow down by. Defaults to 1.
+        hop_len (int): Number audio of frames between STFT columns.
+            Defaults to 512.
+        num_bins (int, optional): number of filter banks from stft.
+            Defaults to 1025.
+    """
+
+    def __init__(self, rate=1., hop_len=512, num_bins=1025):
+
+        super(StretchSpecTime, self).__init__()
+
+        self.rate = rate
+        phi_advance = torch.linspace(
+            0, math.pi * hop_len, num_bins)[..., None]
+
+        self.register_buffer('phi_advance', phi_advance)
+
+    def forward(self, spect, rate=None):
+        if rate is None:
+            rate = self.rate
+        return phase_vocoder(spect, rate, self.phi_advance)
+
+    def __repr__(self):
+        param_str = '(rate={})'.format(self.rate)
+        return self.__class__.__name__ + param_str
+
+
+class GaussianNoise(_ModuleNoStateBuffers):
+    """
+    Apply gaussian noise to a spectrogram.
+
+    Args:
+        scale (float or tensor): standard deviation of the distribution.
+    """
+
+    def __init__(self, scale=1.):
+        super(GaussianNoise, self).__init__()
+        self.scale = torch.as_tensor(scale)
+        self.loc = torch.tensor(0.)
+
+    def forward(self, spect, scale=None):
+        if scale == None:
+            scale = self.scale
+        else:
+            scale = torch.as_tensor(scale)
+
+        with torch.no_grad():
+
+            noise = torch.normal(self.loc.expand(spect.shape), scale.expand(spect.shape)) 
+            return spect, noise
+
+    def __repr__(self):
+        param_str = '(scale={})'.format(self.scale)
+        return self.__class__.__name__ + param_str
+
+
+def apply_masking(spect, max_value, mask_value, axis):
+
+    assert axis in [2,3]
```

Comment:
i recently changed my habit, too, to prefer `raise` to `assert` for some reasons. reference: https://stackoverflow.com/questions/20076182/what-are-the-implications-of-using-assert-vs-raise-exception/20076295